src/main/java/org/candlepin/resource/ConsumerResource.java

```diff
@@ -1675,8 +1682,14 @@ else if (environmentName != null) {
                     this.contentAccessManager.removeContentAccessCert(toUpdate);
                 }
 
-                // lazily regenerate certs, so the client can still work
-                poolManager.regenerateCertificatesOf(toUpdate, true);
+                if (firstPass) {
+                    // lazily regenerate certs, so the client can still work
+                    poolManager.regenerateCertificatesOf(toUpdate, true);
+                }
+                else {
+                    List<String> updatedEnvsList = selectEnvironment(preExistingEnvIds, validatedEnvIds);
+                    poolManager.regenerateCertificatesOf(toUpdate, toUpdate.getEnvironmentIds(), updatedEnvsList, true);
```

Comment:
Without having looked at the PR in detail yet, one thing that I can see is that this whole process is going to be expensive, which is why I suggested to do it using a query. If hard to do with a single query, then two or three of them would be preferable than doing all this in java

Reply:
Yes I understand, I am going to break the query into multiple JPQL queries. I wanted to write the spec test to test out couple of scenarios.